<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Uncover Cell-Cell Interactions in scRNA-seq â€¢ scMatchmaker</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Uncover Cell-Cell Interactions in scRNA-seq">
<meta property="og:description" content="scMatchmaker">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMatchmaker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/example.html">Get started</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/stevexniu/scMatchmaker/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Uncover Cell-Cell Interactions in scRNA-seq</h1>
            
            <h4 data-toc-skip class="date">Last Run: May 14, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stevexniu/scMatchmaker/blob/HEAD/vignettes/example.Rmd" class="external-link"><code>vignettes/example.Rmd</code></a></small>
      <div class="hidden name"><code>example.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="setup-the-matchmaker-object">Setup the Matchmaker Object<a class="anchor" aria-label="anchor" href="#setup-the-matchmaker-object"></a>
</h3>
<p>For this tutorial, we will be analyzing the human decidua single cell RNA-seq dataset from <a href="https://www.nature.com/articles/s41586-018-0698-6" class="external-link">Vento-tormo et al. 2018 Nature</a>. There are 64,734 single cells that were generated using 10X Genomics. The raw data can be found <a href="https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-6701/" class="external-link">here</a>. We will be using the <a href="https://www.cellphonedb.org/" class="external-link">CellPhoneDB</a> ligands-receptors interaction repository.</p>
<p>A fraction of this data containing 947 ligand and receptor genes and 3000 cells (726 EVT, 953 dS2, 788 dNK1, and 533 dM1) is included in the scMatchmaker package called <code>decidua</code> and the corresponding metadata containing cell type annotations called <code>decidua.annotation</code>. This data has been normalized in logTPM (log-transformed transcripts per million), and it is recommended to start with <a href="https://osca.bioconductor.org/normalization.html" class="external-link">normalized data</a>. The <code>Normalization</code> function can be used to normalize raw scRNA-seq counts. It provides widely used <code>logTPM</code> normalization, as well as <code>cosine</code> normalization used in <a href="https://www.nature.com/articles/nbt.4091" class="external-link">Haghverdi et al. 2018 Nature Biotechnology</a>.</p>
<p>Then we load a CellPhoneDB database included in the package called <code>cellphonedb</code>. Alternatively, we can use <code>LoadCellPhoneDB</code> function to load the CellPhoneDB database either locally or from their website.</p>
<p>You can check the <code>Matchmaker</code> object structure and slot information by typing <code><a href="../reference/Matchmaker-class.html">?'Matchmaker-class'</a></code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Install scMatchmaker from CRAN.</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"scMatchmaker"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"scMatchmaker"</span><span class="op">)</span>

<span class="co"># Load scMatchmaker.</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scMatchmaker</span><span class="op">)</span>

<span class="co"># Load the CellPhoneDB interaction database from URLs. </span>
<span class="co"># This step is optional when using scMatchmaker's preloaded database.</span>
<span class="va">interaction.url</span> <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Teichlab/cellphonedb-data/master/data/interaction_input.csv"</span>
<span class="va">gene.url</span> <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Teichlab/cellphonedb-data/master/data/gene_input.csv"</span>
<span class="va">complex.url</span> <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Teichlab/cellphonedb-data/master/data/complex_input.csv"</span> 

<span class="va">interaction.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadCellPhoneDB.html">LoadCellPhoneDB</a></span><span class="op">(</span>
  interaction_input <span class="op">=</span> <span class="va">interaction.url</span>,
  gene_input <span class="op">=</span> <span class="va">gene.url</span>,
  complex_input <span class="op">=</span> <span class="va">complex.url</span>,
  gene.symbol <span class="op">=</span> <span class="st">"gene_name"</span>, url <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Initialize Matchmaker object with scMatchmaker preloaded cellphonedb database.</span>
<span class="co"># Missing ligands and receptors will be added with zeros, this step is useful when calculating protein interaction complex later.</span>
<span class="va">decidua.interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Screening.html">Screening</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">decidua</span>, annotation <span class="op">=</span> <span class="va">decidua.annotation</span><span class="op">[</span>,<span class="st">"annotation"</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, interaction <span class="op">=</span> <span class="va">cellphonedb</span>, project_name <span class="op">=</span> <span class="st">"Decidua"</span><span class="op">)</span>
<span class="va">decidua.interaction</span>
<span class="co">## Decidua Matchmaker object: </span>
<span class="co">## 1574 interaction pairs across 3000 cells. </span>
<span class="co">## 0 ligand-recptor pairs across 0 cell-cell interactions calculated.</span></code></pre></div>
<p><details><summary><strong>What does the data look like?</strong></summary></details></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's look at the some example ligand and receptor genes in the first twenty cells.</span>
<span class="va">decidua</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>,<span class="st">"TGFB1"</span>,<span class="st">"EGF"</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>
<span class="co">## 3 x 20 sparse Matrix of class "dgCMatrix"</span>
<span class="co">##                                                                                                                               </span>
<span class="co">## EGFR  . .        . . .         . . .         . .         . .         .       . .        .        .        . .         .       </span>
<span class="co">## TGFB1 . 1.131673 . . 0.6504162 . . 0.9582769 . 0.7757068 . 0.9077957 1.07289 . 1.122225 1.407436 1.318945 . 0.6545622 1.395667</span>
<span class="co">## EGF   . .        . . .         . . .         . .         . .         .       . .        .        .        . .         .</span></code></pre></div>
<p>The dots <code>.</code> in the decidua dataset count matrix represent 0's (no UMI). Because scRNA-seq data contains lots of 0's, Matchmaker uses this sparse matrix notation to store data, interaction strength and p values matrices to save memory. However, if there are not many zero entries, sparse matrix is no longer efficient to save memory. Matchmaker controls for this by the <code>zero_percent</code> parameter. The default is 0.7, that is, when the non-zero entries is above 70% the data will be saved as sparse matrix,</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's look at the first six rows of the corresponding metadata.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">decidua.annotation</span><span class="op">)</span>
<span class="co">##                             Fetus location final_cluster annotation</span>
<span class="co">## FCA7196224_TCACAAGCAGTTCATG    D9  Decidua             7        dM1</span>
<span class="co">## FCA7167223_AATCCAGTCACGCATA    D6  Decidua             7        dM1</span>
<span class="co">## FCA7167223_GGACAGAAGTCGATAA    D6  Decidua             7        dM1</span>
<span class="co">## FCA7196224_TATCTCAGTTCCCGAG    D9  Decidua             7        dM1</span>
<span class="co">## FCA7196224_GTGCATATCCAATGGT    D9  Decidua             7        dM1</span>
<span class="co">## FCA7167221_CGAATGTAGACCACGA    D7  Decidua             7        dM1</span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's look at the first six rows of the CellPhoneDB interactions loaded from URLs.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">interaction.data</span><span class="op">)</span>
<span class="co">##   genename_a genename_b id_cp_interaction partner_a partner_b protein_name_a protein_name_b annotation_strategy         source subunit_a_1</span>
<span class="co">## 1       ACE2       GHRL   CPI-SS085C150CE    Q9BYF1    Q9UBU3     ACE2_HUMAN     GHRL_HUMAN                 I2D                       &lt;NA&gt;</span>
<span class="co">## 2      ACKR1      CCL17   CPI-SS0974EB232    Q16570    Q92583    ACKR1_HUMAN    CCL17_HUMAN                 I2D                       &lt;NA&gt;</span>
<span class="co">## 3      ACKR2       CCL2   CPI-SS03165AD8C    O00590    P13500    ACKR2_HUMAN     CCL2_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">## 4      ACKR2       CCL3   CPI-SS04318D03E    O00590    P10147    ACKR2_HUMAN     CCL3_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">## 5      ACKR2      CCL14   CPI-SS045A5FC60    O00590    Q16627    ACKR2_HUMAN    CCL14_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">## 6      ACKR2      CCL13   CPI-SS06B0861F1    O00590    Q99616    ACKR2_HUMAN    CCL13_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">##   subunit_a_2 subunit_a_3 subunit_b_1 subunit_b_2 subunit_b_3</span>
<span class="co">## 1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 2        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 3        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 4        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 5        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 6        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's look at the first six rows of the preloaded CellPhoneDB interaction list.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cellphonedb</span><span class="op">)</span>
<span class="co">##   genename_a genename_b id_cp_interaction partner_a partner_b protein_name_a protein_name_b annotation_strategy         source subunit_a_1</span>
<span class="co">## 1       ACE2       GHRL   CPI-SS085C150CE    Q9BYF1    Q9UBU3     ACE2_HUMAN     GHRL_HUMAN                 I2D                       &lt;NA&gt;</span>
<span class="co">## 2      ACKR1      CCL17   CPI-SS0974EB232    Q16570    Q92583    ACKR1_HUMAN    CCL17_HUMAN                 I2D                       &lt;NA&gt;</span>
<span class="co">## 3      ACKR2       CCL2   CPI-SS03165AD8C    O00590    P13500    ACKR2_HUMAN     CCL2_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">## 4      ACKR2       CCL3   CPI-SS04318D03E    O00590    P10147    ACKR2_HUMAN     CCL3_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">## 5      ACKR2      CCL14   CPI-SS045A5FC60    O00590    Q16627    ACKR2_HUMAN    CCL14_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">## 6      ACKR2      CCL13   CPI-SS06B0861F1    O00590    Q99616    ACKR2_HUMAN    CCL13_HUMAN             curated PMID: 24218476        &lt;NA&gt;</span>
<span class="co">##   subunit_a_2 subunit_a_3 subunit_b_1 subunit_b_2 subunit_b_3</span>
<span class="co">## 1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 2        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 3        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 4        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 5        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">## 6        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></code></pre></div>
<p> Â </p>
</div>
<div class="section level3">
<h3 id="identify-cell-cell-interactions-with-matchmaking">Identify Cell-Cell Interactions with Matchmaking<a class="anchor" aria-label="anchor" href="#identify-cell-cell-interactions-with-matchmaking"></a>
</h3>
<p>Before we search for cell-cell interactions, we can optionally downsample the data with <code>Sketching</code> function to reduce the number of cells which may speed up the computation time.</p>
<ul>
<li>There are two methods to perform downsampling:
<ul>
<li>Downsample proportionally to the cell type frequencies (<code>downsampling = TRUE</code>). The actual downsampling size may vary slightly due to rounding effects.</li>
<li>Geometric Sketching by <a href="https://www.cell.com/cell-systems/pdfExtended/S2405-4712(19)30152-8" class="external-link">Hie et al. 2019 Cell Systems</a> using the input data (<code>downsampling = FALSE</code>).</li>
</ul>
</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Downsampling the number of cells to 1000.</span>
<span class="va">decidua.interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Sketching.html">Sketching</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, size <span class="op">=</span> <span class="fl">1000</span>, downsampling <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">decidua.interaction</span>
<span class="co">## Decidua Matchmaker object: </span>
<span class="co">## 1574 interaction pairs across 1001 cells. </span>
<span class="co">## 0 ligand-recptor pairs across 0 cell-cell interactions calculated.</span></code></pre></div>
<p><details><summary><strong>Which cells are being "sketched"?</strong></summary></details></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show the frist six downsampled cell IDs.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">decidua.interaction</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">sketch_id</span><span class="op">)</span>
<span class="co">## [1] 129 509 471 299 270 187</span></code></pre></div>
<p> Â </p>
<p>The <code>Matchmaking</code> function is the workhorse that calculates the interaction strengths and p values. It first calculates the interaction strengths using either a base model or an Earth Mover's Distance (EMD) adjusted model (see below). Then it randomly permutes the cell type identities for a number of times (usually 100-1000), and calculates a null distribution for each interacting pairs similar to the <a href="https://github.com/Teichlab/cellphonedb" class="external-link">CellPhoneDB Python Package</a>. This null distribution will serve as a nonparametric tests to select the significant interactions.</p>
<ul>
<li>
<code>Matchmaking</code> will return two interaction matrices in <code>Matchmaker</code> object:
<ul>
<li>
<code>@strength</code> stores the relative interaction strength.</li>
<li>
<code>@pvalue</code> stores the p values for each interactions.</li>
</ul>
</li>
</ul>
<p>The rows represent cell-cell interacting pairs (<strong>dM1|DC1</strong> stands for interaction between dM1 and DC1 cell types) and the columns represent ligand-receptor pairs (i.e. <strong>ACE2_GHRL</strong> stands for interaction between ACE2 and GHRL genes). Together it reads as <strong>dM1</strong> cells express <strong>ACE2</strong> molecules and interact with <strong>GHRL</strong> molecules on <strong>DC1</strong> cells. Please note that ligand receptor interactions are direction sensitive, <strong>dM1|DC1</strong> has opposite in meaning to <strong>DC1|dM1</strong>.</p>
<ul>
<li>
<code>Matchmaking</code> has two methods to calculate interaction strength:
<ul>
<li>Default base model (<code>emd = FALSE</code>). Interaction strengths are calculated as average mean expressions between interacting ligand and receptor pairs.</li>
<li>Earth Mover's Distance (EMD) adjusted (<code>emd = TRUE</code>). Interaction strengths are first calculated as average mean expressions the same as the base model, and then adjusted by EMD which takes into account the similarity between ligand and receptor distributions. If the two distributions are similar to each other, the EMD will be small, vice versa.</li>
<li>The EMD model has two modes: weighted (<code>weighted = TRUE</code>) or unweighted (<code>weighted = FALSE</code>). The weighted model will assign inversely proportional to the expression level, thus, higher. The unweighted model on the other hand will assign equal weights to all the expression levels.</li>
</ul>
</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Base model with 100 permutations.</span>
<span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">decidua.interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Matchmaking.html">Matchmaking</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, n_perm <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span>
<span class="co">##    user  system elapsed </span>
<span class="co">##   1.455   0.174  24.130</span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">decidua.interaction</span>
<span class="co">## Decidua Matchmaker object: </span>
<span class="co">## 1574 interaction pairs across 1001 cells. </span>
<span class="co">## 1574 ligand-recptor pairs across 16 cell-cell interactions calculated.</span></code></pre></div>
<p>We run the unweighted EMD model with 100 permutations.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># It will take longer time compared to the base model because of the additional EMD adjustment.</span>
<span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">decidua.interaction.emd.unweighted</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Matchmaking.html">Matchmaking</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, emd <span class="op">=</span> <span class="cn">TRUE</span>, n_perm <span class="op">=</span> <span class="fl">100</span>, weighted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span>
<span class="co">##    user  system elapsed </span>
<span class="co">##   4.669   0.421 273.400</span></code></pre></div>
<p>We run the default weighted EMD model with 100 permutations.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># It will take longer time compared to the unweighted model.</span>
<span class="va">ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">decidua.interaction.emd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Matchmaking.html">Matchmaking</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, emd <span class="op">=</span> <span class="cn">TRUE</span>, n_perm <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ptm</span>
<span class="co">##    user  system elapsed </span>
<span class="co">##   8.785   0.551 267.769</span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">decidua.interaction.emd</span>
<span class="co">## Decidua Matchmaker object: </span>
<span class="co">## 1574 interaction pairs across 1001 cells. </span>
<span class="co">## 1574 ligand-recptor pairs across 16 cell-cell interactions calculated.</span></code></pre></div>
<ul>
<li>Key arguments passed to <code>Matchmaking</code> function include:
<ul>
<li>
<code>n_perm</code> Number of random permutations. Default is <code>100</code>.</li>
<li>
<code>nbins</code> Number of bins to use for Earth Mover's Distance calculation.</li>
<li>
<code>p.adjust.method</code> p value adjustment method. Default is <code>"BH"</code> for Benjamini &amp; Hochberg method.</li>
<li>
<code>emd</code> Whether to run EMD model. Default is FALSE</li>
<li>
<code>weighted</code> Whether to run weighted EMD model. Default is TRUE.</li>
<li>Additional arguments can be found by typing <code><a href="../reference/Matchmaking.html">?Matchmaking</a></code>.</li>
</ul>
</li>
</ul>
<p><details><summary><strong>What do the interaction strength and p values matrices look like?</strong></summary></details></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show the first five rows and the first five columns of the strength matrix.</span>
<span class="va">decidua.interaction.emd</span><span class="op">@</span><span class="va">strength</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">##           ACE2_GHRL ACKR1_CCL17 ACKR2_CCL2 ACKR2_CCL3 ACKR2_CCL14</span>
<span class="co">## dM1|dM1  0.00000000 0.000000000  0.0000000  0.0000000           0</span>
<span class="co">## dNK1|dM1 0.00000000 0.000000000  0.0000000  0.0000000           0</span>
<span class="co">## dS2|dM1  0.01649279 0.015862922  0.2669892  0.4659861           0</span>
<span class="co">## EVT|dM1  0.01257415 0.009649865  0.4835376  0.7245980           0</span>
<span class="co">## dM1|dNK1 0.00000000 0.000000000  0.0000000  0.0000000           0</span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show the first five rows and the first five columns of the p value matrix.</span>
<span class="va">decidua.interaction.emd</span><span class="op">@</span><span class="va">pvalue</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">##           ACE2_GHRL ACKR1_CCL17 ACKR2_CCL2 ACKR2_CCL3 ACKR2_CCL14</span>
<span class="co">## dM1|dM1  1.00000000  1.00000000 1.00000000 1.00000000           1</span>
<span class="co">## dNK1|dM1 1.00000000  1.00000000 1.00000000 1.00000000           1</span>
<span class="co">## dS2|dM1  0.05940594  0.00990099 0.00990099 1.00000000           1</span>
<span class="co">## EVT|dM1  0.17821782  0.04950495 0.00990099 0.00990099           1</span>
<span class="co">## dM1|dNK1 1.00000000  1.00000000 1.00000000 1.00000000           1</span></code></pre></div>
<p> Â </p>
<p>If interaction complexes information is provided in the <code>@interaction</code> slot, for example, in the preloaded <code>cellphonedb</code> database, the subunits are listed with column names <code>subunit_a_1</code>, <code>subunit_a_2</code>, <code>subunit_a_3</code> and <code>subunit_b_1</code>, <code>subunit_b_2</code>, <code>subunit_b_3</code> respectively.</p>
<p>We can calculate the complex-complex interactions by using <code>Complexing</code> function. It calculates the complex interactions from its subunits with the following options:</p>
<ul>
<li>
<code>strength_comb</code> Method to combine subunits strengths.
<ul>
<li>
<code>min</code>, minimum strength of its subunits (Default).</li>
<li>
<code>average</code>, average strength of its subunits.</li>
<li>
<code>max</code>, maximum strength of its subunits.</li>
</ul>
</li>
<li>
<code>pval_comb</code> Method to combine subunits p values.
<ul>
<li>
<code>max</code>, maximum p value of its subunits (Default).</li>
<li>
<code>average</code>, average p value of its subunits.</li>
<li>
<code>min</code>, minimum p value of its subunits.</li>
</ul>
</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Please note that if Selecting function is called before Complexing, you will need to re-run the Selecting step.</span>
<span class="va">decidua.interaction.emd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Complexing.html">Complexing</a></span><span class="op">(</span><span class="va">decidua.interaction.emd</span><span class="op">)</span></code></pre></div>
<p>We can also merge directed (one-way) interactions into undirected (two-way) interactionsc using <code>Merging</code> function. For example, before merging, <em>DC1|dM1</em> and <em>dM1|DC1</em> represents two different cell-cell interactions: DC1 expresses ligands and dM1 expresses receptors, versus, dM1 expresses ligands and DC1 expresses receptors. After merging, they will be combined as one undirected cell-cell interaction between DC1 and dM1.</p>
<p>The <code>Merging</code> function takes following arguments similar to <code>Complexing</code>:</p>
<ul>
<li>
<code>strength_merge</code> Method to merge directed interaction strengths.
<ul>
<li>
<code>max</code>, maximum strength of directed interactions (Default).</li>
<li>
<code>average</code>, average strength of directed interactions.</li>
<li>
<code>min</code>, minimum strength of directed interaction.</li>
</ul>
</li>
<li>
<code>pval_merge</code> Method to directed interactions p values.
<ul>
<li>
<code>min</code>, minimum p value of directed interactions (Default).</li>
<li>
<code>average</code>, average p value of directed interactions.</li>
<li>
<code>max</code>, maximum p value of directed interactions.</li>
</ul>
</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Merge the directed one-way interactions into the undirected two-way interactions.</span>
<span class="va">decidua.interaction.emd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Merging.html">Merging</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span><span class="op">)</span></code></pre></div>
<p>If we want to revert the <code>Complexing</code> or <code>Merging</code> operation, we can use the <code>Resetting</code> function by specifying the <code>by</code> argument with either <code>complex</code> (default) or <code>merge</code> to revert it.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reset the Merging operation. </span>
<span class="va">decidua.interaction.emd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Resetting.html">Resetting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, by <span class="op">=</span> <span class="st">"merge"</span><span class="op">)</span>

<span class="co"># Reset the Complexing operation. </span>
<span class="va">decidua.interaction.emd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Resetting.html">Resetting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, by <span class="op">=</span> <span class="st">"complex"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-top-interactions">Select Top Interactions<a class="anchor" aria-label="anchor" href="#select-top-interactions"></a>
</h3>
<p>After calculating the interaction strengths and p values. We use <code>Selecting</code> function to filter and select the top significant interactions.</p>
<ul>
<li>Aguments passed to <code>Selecting</code> include:
<ul>
<li>The <code>strength.pct</code> argument defines the quantile of interaction strength to be selected. The default is 0.1 (top 10% will be chosen).</li>
<li>The <code>pval.cutoff</code> argument defines the p value cutoff, The default is 0.05 (interaction with p value less than 0.05 will be selected).</li>
</ul>
</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Select the top 10% interactions with p values less than 0.05 in the base model.</span>
<span class="va">decidua.interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Selecting.html">Selecting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, strength.pct <span class="op">=</span> <span class="fl">0.1</span>, pval.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>

<span class="co"># Select the top 10% interactions with p values less than 0.05 in the EMD model.</span>
<span class="va">decidua.interaction.emd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Selecting.html">Selecting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, strength.pct <span class="op">=</span> <span class="fl">0.1</span>, pval.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<p>The <code>Converting</code> function converts the strength and p value matrices into a long ranked list of interaction candidates. By setting <code>selected = TRUE</code>, it will convert the selected data from the <code>Selecting</code> step.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Convert the selected interactions in base model.</span>
<span class="va">converted.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Converting.html">Converting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, selected <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Convert the selected interactions in EMD model.</span>
<span class="va">converted.data.emd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Converting.html">Converting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, selected <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>We can subset the cell-cell interactions or ligand-receptor pairs using <code>Subseting</code> function.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Subset interactions between EVT and dM1.</span>
<span class="va">dc1.m1.interactions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Subsetting.html">Subsetting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, ident1 <span class="op">=</span> <span class="st">"EVT"</span>, ident2 <span class="op">=</span> <span class="st">"dM1"</span><span class="op">)</span>

<span class="co"># Subset cell-cell interactions involving PGF-FLT1.</span>
<span class="va">pgf.flt1.interactions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Subsetting.html">Subsetting</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, partner_a <span class="op">=</span> <span class="st">"FLT1"</span>, partner_b <span class="op">=</span> <span class="st">"PGF"</span><span class="op">)</span></code></pre></div>
<p>Finally, we can save the significant (selected) interaction strengths and p-values into csv files using <code>Saving</code> function.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Save the results.</span>
<span class="fu"><a href="../reference/Saving.html">Saving</a></span><span class="op">(</span><span class="va">decidua.interaction.emd</span>, file_name <span class="op">=</span> <span class="st">"decidua_emd"</span>, selected <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>We can also save the analyzed Matchmaker R object.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Save the Matchmaker object.</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, file <span class="op">=</span> <span class="st">"decidua.interaction.emd.rds"</span><span class="op">)</span></code></pre></div>
<p><details><summary><strong>What are the top selected interactions?</strong></summary></details></p>
<p>The top ten interactions in the base model:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show the top ten interactions in the base model.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">converted.data</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">##    interactions       pairs strength       pval</span>
<span class="co">## 1       dM1|EVT    CD74_MIF 3.003241 0.00990099</span>
<span class="co">## 2      dM1|dNK1 C5AR1_RPS19 2.894307 0.00990099</span>
<span class="co">## 3       dM1|dM1    CD74_MIF 2.779309 0.00990099</span>
<span class="co">## 4      dM1|dNK1    CD74_MIF 2.739418 0.00990099</span>
<span class="co">## 5       dM1|dS2    CD74_MIF 2.647261 0.00990099</span>
<span class="co">## 6       dM1|dS2    CD74_APP 2.609492 0.00990099</span>
<span class="co">## 7       dM1|EVT C5AR1_RPS19 2.607603 0.00990099</span>
<span class="co">## 8      dNK1|EVT     B2M_ALB 2.535686 0.00990099</span>
<span class="co">## 9      dNK1|dS2     B2M_ALB 2.509800 0.00990099</span>
<span class="co">## 10      dM1|EVT     B2M_ALB 2.457247 0.00990099</span></code></pre></div>
<p>The top ten interactions in the EMD model:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show the top ten interactions in the EMD model.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">converted.data.emd</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">##    interactions        pairs strength       pval</span>
<span class="co">## 1     dNK1|dNK1  KLRC1_HLA-E 2.117210 0.00990099</span>
<span class="co">## 2      dNK1|dM1  KLRC1_HLA-E 1.798782 0.00990099</span>
<span class="co">## 3       EVT|EVT     FLT1_PGF 1.744274 0.00990099</span>
<span class="co">## 4       dM1|EVT     CD74_MIF 1.744064 0.00990099</span>
<span class="co">## 5       EVT|dS2    FN1_ITGB1 1.668484 0.00990099</span>
<span class="co">## 6       dS2|dS2 COL6A2_ITGB1 1.634794 0.00990099</span>
<span class="co">## 7       EVT|EVT    FN1_ITGB1 1.629702 0.00990099</span>
<span class="co">## 8       dM1|dS2  PLAUR_ITGB1 1.629599 0.00990099</span>
<span class="co">## 9       dM1|dM1    SPP1_CD44 1.606700 0.00990099</span>
<span class="co">## 10    dNK1|dNK1  KLRD1_HLA-E 1.595239 0.00990099</span></code></pre></div>
<p> Â </p>
</div>
<div class="section level3">
<h3 id="visualize-cell-cell-interactions">Visualize Cell-Cell Interactions<a class="anchor" aria-label="anchor" href="#visualize-cell-cell-interactions"></a>
</h3>
<p><code>scMatchmaker</code> provides three different ways to visualize the interactions:<br><code>PlotSpot</code> function generates a spot plot either with bars or dots.<br><code>PlotHeatmap</code> function generates a heatmap plot. <code>PlotNetwork</code> function generates a network plot. <code>PlotScatter</code> function generates a scatter plot for a specific cell-cell interaction pair.<br><code>PlotHistogram</code> function generates a histogram for a specific ligand and receptor pair.</p>
<ul>
<li>
<code>PlotSpot</code> generates a spot plot for cell-cell interactions. The size or height of the dot or bar represents the interaction strength. The color of the dot or bar represents the p values. Some useful arguments are listed below:
<ul>
<li>
<code>which.cell</code> and <code>which.gene</code> cell-cell or gene-gene interaction name or index to plot.</li>
<li>
<code>type</code> option <code>"dot"</code> for dot plot and option <code>"bar"</code> for bar plot.</li>
<li>
<code>order</code> order the data.</li>
<li>
<code>low.cutoff</code> and <code>high.cutoff</code> Cutoff for values to show.</li>
<li>
<code>plot.y.axis</code>, <code>fix.y.axis</code> etc. See <code><a href="../reference/PlotSpot.html">?PlotSpot</a></code> for details.</li>
</ul>
</li>
</ul>
<p>We use <code>PlotSpot</code> to visualize the first twenty rows and the first ten columns of the interaction <code>@strength</code> and <code>@pvalues</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Spot plot with bars </span>
<span class="va">plot.bar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotSpot.html">PlotSpot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, which.cell <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, which.gene <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, type <span class="op">=</span> <span class="st">"bar"</span>, order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Spot plot with dots</span>
<span class="va">plot.dot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotSpot.html">PlotSpot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, which.cell <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, which.gene <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, type <span class="op">=</span> <span class="st">"dot"</span>, order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">plot.bar</span>, <span class="va">plot.dot</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/spotplot-1.png" width="1344" style="display: block; margin: auto;"></p>
<ul>
<li>
<code>PlotHeatmap</code> generates a heatmap. It takes arguments similar to <code>PlotSpot</code>. In addition, it can provide a summarized view by aggregating interaction data:
<ul>
<li>
<code>selected</code> use <code>@selected</code> data if calculated.</li>
<li>
<code>aggregate</code> type of aggregation to use. Options include <code>cell</code> and <code>gene</code>.</li>
<li>
<code>counted</code> aggregate interaction counts instead of strengths.</li>
<li>
<code>scale</code> scale and center the data to plot.</li>
<li>Additional arguments see <code><a href="../reference/PlotHeatmap.html">?PlotHeatmap</a></code> for details.</li>
</ul>
</li>
</ul>
<p>We use <code>PlotHeatmap</code> to visualize the most interactive cell types by setting <code>aggregate = "cell"</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/PlotHeatmap.html">PlotHeatmap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, aggregate <span class="op">=</span> <span class="st">"cell"</span>, order <span class="op">=</span> <span class="cn">TRUE</span>, selected <span class="op">=</span> <span class="cn">TRUE</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/heatmap-1.png" width="672" style="display: block; margin: auto;"></p>
<ul>
<li>
<code>PlotNetwork</code> generates a gene-gene or cell-cell interaction network plot. It takes arguments similar to <code>PlotHeatmap</code>:
<ul>
<li>
<code>selected</code> use <code>@selected</code> data if calculated.</li>
<li>
<code>aggregate</code> type of aggregation to use. Options include <code>cell</code> and <code>gene</code>.</li>
<li>
<code>node.size</code>, <code>node.color</code> and <code>edge.color</code> size and color of the nodes and edges.</li>
<li>
<code>layout</code> layout used to plot network (i.e tsne or umap layout).</li>
<li>
<code>directed</code> plots directed network.</li>
<li>
<code>louvain</code> performs louvain clustering.</li>
<li>Additional arguments see <code><a href="../reference/PlotNetwork.html">?PlotNetwork</a></code> for details.</li>
</ul>
</li>
</ul>
<p>We use <code>PlotNetwork</code> to visualize the cell-cell interaction network by setting <code>aggregate = "cell"</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/PlotNetwork.html">PlotNetwork</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, aggregate <span class="op">=</span> <span class="st">"cell"</span>, selected <span class="op">=</span> <span class="cn">TRUE</span>, legend.posit <span class="op">=</span> <span class="st">"topleft"</span>, legend.breaks <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/network.cell-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can also use <code>PlotNetwork</code> to visualize the gene-gene interaction network by setting <code>aggregate = "gene"</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/PlotNetwork.html">PlotNetwork</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, aggregate <span class="op">=</span> <span class="st">"gene"</span>, louvain <span class="op">=</span> <span class="cn">TRUE</span>, selected <span class="op">=</span> <span class="cn">TRUE</span>, filter <span class="op">=</span> <span class="cn">TRUE</span>, low.cutoff <span class="op">=</span> <span class="fl">5</span>, node.size <span class="op">=</span> <span class="fl">4</span>, legend.posit <span class="op">=</span> <span class="st">"bottomright"</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/network.gene-1.png" width="672" style="display: block; margin: auto;"></p>
<ul>
<li>The <code>PlotScatter</code> function generates visualization of expression profiles for specific gene-gene interactions among all other interactions in a given pair of cell-cell interaction.</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Visualize ITGB1-PLAUR interactions between EVT and dM1 cells.</span>
<span class="fu"><a href="../reference/PlotScatter.html">PlotScatter</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, ident1 <span class="op">=</span> <span class="st">"dM1"</span>, ident2 <span class="op">=</span> <span class="st">"EVT"</span>, ligands <span class="op">=</span> <span class="st">"ITGB1"</span>, receptors <span class="op">=</span> <span class="st">"PLAUR"</span>, add.lines <span class="op">=</span> <span class="cn">TRUE</span>, point.cex <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/scatterplot-1.png" width="672" style="display: block; margin: auto;"></p>
<ul>
<li>
<code>PlotHistogram</code> generates a histogram to visualize the ligand and receptor expressions in interacting cell types. Possible arguments include:
<ul>
<li>
<code>ligand</code> and <code>receptor</code> ligand and receptor gene names.</li>
<li>
<code>ligand.ident</code> and <code>receptor.ident</code> Cell types to plot for ligand and receptor.</li>
<li>
<code>nbins</code> Number of bins to use for histogram.</li>
<li>Additional arguments see <code><a href="../reference/PlotHistogram.html">?PlotHistogram</a></code> for details.</li>
</ul>
</li>
</ul>
<p>We use <code>PlotHistogram</code> to visualize the top interacting ligand-receptor pairs in the base model: <em>CD74_MIF</em> and <em>CD74_APP</em> in <em>dM1|EVT</em> and <em>dM1|dS2</em> cell-cell interactions respectively. We can see that the base model prioritizes higher mean expression values of ligand and receptor pairs, regardless of their variable distributions.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Histogram for CD74 and MIF interaction between dM1 and EVT cells.</span>
<span class="fu"><a href="../reference/PlotHistogram.html">PlotHistogram</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, ligand <span class="op">=</span> <span class="st">"MIF"</span>, receptor <span class="op">=</span> <span class="st">"CD74"</span>, ligand.ident <span class="op">=</span> <span class="st">"EVT"</span>, receptor.ident <span class="op">=</span> <span class="st">"dM1"</span>, nbins <span class="op">=</span> <span class="fl">25</span><span class="op">)</span>

<span class="co"># Histogram for CD74 and APP interaction between dM1 and dS2 cells.</span>
<span class="fu"><a href="../reference/PlotHistogram.html">PlotHistogram</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction</span>, ligand <span class="op">=</span> <span class="st">"APP"</span>, receptor <span class="op">=</span> <span class="st">"CD74"</span>, ligand.ident <span class="op">=</span> <span class="st">"dS2"</span>, receptor.ident <span class="op">=</span> <span class="st">"dM1"</span>, nbins <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/histogram.base-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We then use <code>PlotHistogram</code> to visualize the top interacting ligand-receptor pairs in the EMD model: <em>FLT1_PGF</em> and <em>KLRC1_HLA-E</em> in <em>EVT|EVT</em> and <em>dNK1|dM1</em> cell-cell interactions respectively. The results show that the EMD model takes into account the variability in the distributions between ligand and receptor pair. It is not biased towards higher mean expression values in either ligand or receptor. It selects the best "matching" pairs.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Histogram for KLRC1 and HLA-E interaction between dNK1 and dM1 cells.</span>
<span class="fu"><a href="../reference/PlotHistogram.html">PlotHistogram</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, ligand <span class="op">=</span> <span class="st">"KLRC1"</span>, receptor <span class="op">=</span> <span class="st">"HLA-E"</span>, ligand.ident <span class="op">=</span> <span class="st">"dNK1"</span>, receptor.ident <span class="op">=</span> <span class="st">"dM1"</span>, nbins <span class="op">=</span> <span class="fl">25</span><span class="op">)</span>

<span class="co"># Histogram for FLT1 and PGF interaction between EVT and EVT cells.</span>
<span class="fu"><a href="../reference/PlotHistogram.html">PlotHistogram</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">decidua.interaction.emd</span>, ligand <span class="op">=</span> <span class="st">"PGF"</span>, receptor <span class="op">=</span> <span class="st">"FLT1"</span>, ligand.ident <span class="op">=</span> <span class="st">"EVT"</span>, receptor.ident <span class="op">=</span> <span class="st">"EVT"</span>, nbins <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/histogram.emd-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Lastly, we compare the base model (x axis) against the EMD model (y axis). We see that the baseline model creates quite a few false-positives (dark red dots in lower right).</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">decidua.interaction</span><span class="op">@</span><span class="va">strength</span>, 
     y <span class="op">=</span> <span class="va">decidua.interaction.emd</span><span class="op">@</span><span class="va">strength</span>, 
     xlab <span class="op">=</span> <span class="st">"Base Model"</span>, ylab <span class="op">=</span> <span class="st">"EMD Adjusted"</span>, 
     col <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">5</span>,<span class="st">"Reds"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">decidua.interaction</span><span class="op">@</span><span class="va">strength</span><span class="op">-</span><span class="va">decidua.interaction.emd</span><span class="op">@</span><span class="va">strength</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span>,
     pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.2</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">2</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">2</span>,lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="example_files/figure-html/comparison-1.png" width="576" style="display: block; margin: auto;"></p>
<hr>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># R session information.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## R version 4.0.3 (2020-10-10)</span>
<span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">## Running under: macOS Big Sur 10.16</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] scMatchmaker_0.99.0</span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] nlme_3.1-151        bitops_1.0-6        matrixStats_0.57.0  fs_1.5.0            doParallel_1.0.16   RColorBrewer_1.1-2  rprojroot_2.0.2    </span>
<span class="co">##   [8] tools_4.0.3         backports_1.2.1     bslib_0.3.1         R6_2.5.0            irlba_2.3.3         d3Network_0.5.2.1   rpart_4.1-15       </span>
<span class="co">##  [15] KernSmooth_2.23-18  Hmisc_4.4-2         colorspace_2.0-0    nnet_7.3-14         tidyselect_1.1.0    gridExtra_2.3       mnormt_2.0.2       </span>
<span class="co">##  [22] compiler_4.0.3      qgraph_1.6.5        fdrtool_1.2.16      textshaping_0.2.1   cli_3.2.0           htmlTable_2.1.0     TSP_1.1-10         </span>
<span class="co">##  [29] desc_1.4.1          labeling_0.4.2      sass_0.4.0          caTools_1.18.0      scales_1.1.1        checkmate_2.0.0     psych_2.0.12       </span>
<span class="co">##  [36] pbapply_1.4-3       pkgdown_2.0.2       pbivnorm_0.6.0      systemfonts_0.3.2   stringr_1.4.0       digest_0.6.27       foreign_0.8-81     </span>
<span class="co">##  [43] rmarkdown_2.13      base64enc_0.1-3     jpeg_0.1-8.1        pkgconfig_2.0.3     htmltools_0.5.2     highr_0.8           fastmap_1.1.0      </span>
<span class="co">##  [50] htmlwidgets_1.5.3   rlang_1.0.2         huge_1.3.4.1        rstudioapi_0.13     farver_2.0.3        jquerylib_0.1.3     generics_0.1.0     </span>
<span class="co">##  [57] jsonlite_1.7.2      gtools_3.8.2        dplyr_1.0.2         magrittr_2.0.1      Formula_1.2-4       Matrix_1.3-2        Rcpp_1.0.7         </span>
<span class="co">##  [64] munsell_0.5.0       reticulate_1.18     abind_1.4-5         lifecycle_1.0.1     whisker_0.4         stringi_1.5.3       yaml_2.2.1         </span>
<span class="co">##  [71] MASS_7.3-53         gplots_3.1.1        plyr_1.8.6          lavaan_0.6-7        grid_4.0.3          parallel_4.0.3      crayon_1.3.4       </span>
<span class="co">##  [78] lattice_0.20-41     cowplot_1.1.1       splines_4.0.3       tmvnsim_1.0-2       knitr_1.38          pillar_1.4.7        igraph_1.2.6       </span>
<span class="co">##  [85] rjson_0.2.20        corpcor_1.6.9       BDgraph_2.63        stats4_4.0.3        reshape2_1.4.4      codetools_0.2-18    glue_1.6.2         </span>
<span class="co">##  [92] evaluate_0.15       emdist_0.3-1        latticeExtra_0.6-29 data.table_1.13.6   png_0.1-7           vctrs_0.4.1         foreach_1.5.1      </span>
<span class="co">##  [99] gtable_0.3.0        purrr_0.3.4         cachem_1.0.1        ggplot2_3.3.3       xfun_0.30           glasso_1.11         ragg_0.4.0         </span>
<span class="co">## [106] survival_3.2-7      seriation_1.2-9     tibble_3.0.4        iterators_1.0.13    registry_0.5-1      memoise_2.0.1       cluster_2.1.0      </span>
<span class="co">## [113] ellipsis_0.3.2</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by X. Steve Niu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
